File: Golang/test/valid_palindrom.go
package test

func ValidPalindrome(s string) bool { // kasur rusak
	left := 0 // pointer kiri
	right := len(s) - 1 // pointer kanan

	for left < right { // pastiin left selalu lebih kecil dari right biar ga cross gitu nanti
		for left < right && !isAlpanumeric(s[left]) { 
			left ++ // kita longkap kalo ada karakter yang bukan alfanumerik
		}

		for left < right && !isAlpanumeric(s[right]) {
			right --
		}

		if toLower(s[left]) != toLower(s[right]) { // kita bikin tiap karakter lowercase dulu baru kita bandingin
			return false
		}

		left++ // geser dari kiri kekanan
		right-- // geser dari kanan kekiri
	}

	return true
}

func isAlpanumeric(c byte) bool { // tiap karakter itu adalah byte kalo di go
	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')
}

func toLower(c byte) byte {
	if c >= 'A' && c <= 'Z' {
		return c + 32 	// biar dia lowercase
	}

	return c
}

File: Javascript/day_3_hashmap_and_two_pointer/validPalindrom.js
function validPalindrome(words) {
  let left = 0;
  let right = words.length - 1;

  while (left < right) {
    while(left < right && !isAlpanumeric(words[left])) {
      left ++
    }

    while(left < right && !isAlpanumeric(words[right])) {
      right --
    }

    if (words[left].toLowerCase() !== words[right].toLowerCase()) {
      return false
    }

    left++
    right--
  }

  return true
}

console.time()
console.log(validPalindrome("kasur rusak"))
console.log(validPalindrome("bahlil bangsat"))
console.timeEnd()

function isAlpanumeric(char) {
  return /^[a-z0-9]$/i.test(char)
}
